### PR TITLE
Use blivet's "device ID" as a unique device identifier

### PR DESCRIPTION
**Pull request by vojtechtrefny**
Anaconda currently uses name to identify devices internally which causes a lot of problems because multiple devices can share the same name. To fix this blivet introduced a new "device ID" which is unique even for devices with the same name (for technologies that support multiple devices with the same name and for devices sharing the same name but using a different storage technology).

-----

This is not yet finished, but it should be usable for testing and further development on the WebUI side. The blivet change (https://github.com/storaged-project/blivet/pull/1182) is merged but not yet released. Unit tests should be passing on this, I also tried running manual installation with mount point assignment in TUI and also running (most of) kickstart tests and everything seems to be working so far. The Gtk custom partitioning is not yet adjusted to this change but is somewhat working if you ignore the weird "names" displayed in the UI.

### CODE REVIEW

**Review by M4rtinK — APPROVED** [review]
Looks good to me code wise, but I'm afraid I can't really judge the full context of this change in regards to Blivet & Web UI. :P

**Comment by rvykydal** [comment]
Looks good to me.
The Web UI works fine with the only difference in showing device.device-id instead of device.name. We need to accomodate the tests and show a better device description than the original device-id.

**Comment by rvykydal** [comment]
If I create biosboot partition, boot partition and btrfs volume with a subvolume in terminal:
```
set -x

DISK=$1

sgdisk --zap-all ${DISK}
sgdisk --new=0:0:+1MiB --typecode=0:ef02 ${DISK}
sgdisk --new=0:0:+1GB ${DISK}
mkfs.ext4 -F ${DISK}2
sgdisk --new=0:0:+10GiB ${DISK}
mkfs.btrfs -f -L v3 ${DISK}3

# DUP
#sgdisk --new=0:0:+1GiB ${DISK}
#mkfs.btrfs -f -L v4 ${DISK}4


##mkfs.btrfs -f -f -L btrfstest ${DISK}3
##sgdisk --new=0:0:+2GiB ${DISK}
##mkfs.btrfs -f -f -L btrfstest ${DISK}4

TMP_MOUNT="/tmp/btrfs-mount-test"
mkdir -p ${TMP_MOUNT}

mount ${DISK}3 ${TMP_MOUNT}
btrfs subvolume create ${TMP_MOUNT}/root
umount ${TMP_MOUNT}

# DUP
#mount ${DISK}4 ${TMP_MOUNT}
#btrfs subvolume create ${TMP_MOUNT}/root
#umount ${TMP_MOUNT}

rmdir ${TMP_MOUNT}


#sgdisk --new=0:0:0 ${DISK}
#mkfs.xfs -f ${DISK}4


udevadm trigger
udevadm settle --timeout=120

lsblk
blkid
```
Then rescan the storage, go to mount points assignment, assign the partitions with reformatting I get failure in:
```
INFO:blivet:registered action: [224] create device btrfs subvolume root (id 219)
DEBUG:blivet:             DeviceTree.get_device_by_device_id: device_id: BTRFS-dafddd9d-7a42-414d-b66c-223d290d6b03-root ; incomplete: False ; hidden: False ;
DEBUG:blivet:             DeviceTree.get_device_by_device_id returned None
INFO:anaconda.core.threads:Thread Failed: AnaTaskThread-ManualPartitioningTask-1 (139822322681536)
ERROR:anaconda.modules.common.task.task:Thread AnaTaskThread-ManualPartitioningTask-1 has failed: Traceback (most recent call last):
  File "/usr/lib64/python3.12/site-packages/pyanaconda/core/threads.py", line 280, in run
    threading.Thread.run(self)
  File "/usr/lib64/python3.12/threading.py", line 1010, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib64/python3.12/site-packages/pyanaconda/modules/common/task/task.py", line 94, in _thread_run_callback
    self._task_run_callback()
  File "/usr/lib64/python3.12/site-packages/pyanaconda/modules/common/task/task.py", line 107, in _task_run_callback
    self._set_result(self.run())
                     ^^^^^^^^^^
  File "/usr/lib64/python3.12/site-packages/pyanaconda/modules/storage/partitioning/base_partitioning.py", line 53, in run
    self._run(self._storage)
  File "/usr/lib64/python3.12/site-packages/pyanaconda/modules/storage/partitioning/automatic/noninteractive_partitioning.py", line 43, in _run
    self._configure_partitioning(storage)
  File "/usr/lib64/python3.12/site-packages/pyanaconda/modules/storage/partitioning/manual/manual_partitioning.py", line 53, in _configure_partitioning
    self._setup_mount_point(storage, mount_data)
  File "/usr/lib64/python3.12/site-packages/pyanaconda/modules/storage/partitioning/manual/manual_partitioning.py", line 116, in _setup_mount_point
    mount_data.mount_options = device.format.options
                               ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'format'

```

[storage.log](https://github.com/rhinstaller/anaconda/files/14258636/storage.log)

I think it happens because the recreated subvolume is not found here:
https://github.com/vojtechtrefny/anaconda/blob/8dba5f232bfc90141a16859fe18f465f7a822345/pyanaconda/modules/storage/partitioning/manual/manual_partitioning.py#L169

It used to work (at least got past this point as I checked) but I am not 100% sure my use case makes sense. (Maybe the issue is that the volume is recreated as well because the only subvolume is destroyed so the uuids are new?)
I am looking into it.

**Comment by rvykydal** [comment]
> It used to work (at least got past this point as I checked) but I am not 100% sure my use case makes sense. (Maybe the issue is that the volume is recreated as well because the only subvolume is destroyed so the uuids are new?) I am looking into it.

Yes, the issue here seems that if the volume is destroyed together with the subvolume and then recreated, finding the subvolume by device_id containing uuid of the former subvolume fails.

**Comment by vojtechtrefny** [comment]
@rvykydal I've added a new commit that should fix this issue.

**Comment by rvykydal** [comment]
> @rvykydal I've added a new commit that should fix this issue.

Thank you, I have just a few nitpicks.

File: pyanaconda/modules/storage/partitioning/manual/manual_partitioning.py [hunk 1]
```diff
@@ -150,22 +150,39 @@ def _recreate_btrfs_volume(self, storage, device):
         storage.create_device(new_btrfs)
         return new_btrfs
 
+    def _recreate_btrfs_subvolume(self, storage, device):
+        """Recreate a btrfs subvolume device by destroying and adding it.
+
+        :param storage: an instance of the Blivet's storage object
+        :param dev_spec: a BtrfsSubVolumeDevice to recreate
```
Comment (rvykydal) [inline pyanaconda/modules/storage/partitioning/manual/manual_partitioning.py:157]: the param is `device`, please update it also for `_recreate_btrfs_volume` when we are on it.

**Comment by vojtechtrefny** [comment]
This is mostly finished now. I plan to do some more manual testing to be sure everything works, but other than that I don't plan to make more changes.

If you wish to help with testing, you can find updates image with all the changes (and latest blivet which also has some fixes) is available [here](https://vtrefny.fedorapeople.org/img/deviceid.img).

File: pyanaconda/ui/gui/spokes/lib/resize.glade [hunk 2]
```diff
@@ -26,78 +28,78 @@
   </object>
   <object class="GtkAdjustment" id="resizeAdjustment">
     <property name="upper">100</property>
-    <property name="step_increment">1</property>
```
Comment (KKoukiou) [inline pyanaconda/ui/gui/spokes/lib/resize.glade:29]: These changes are unrelated. I guess these are autogenerated. Is there a reason for storing these here?

**Review by KKoukiou — COMMENTED** [review]
Should be mark in the specfile tha tthis current anaconda-core is incomptatible with the current version of the Web UI package?

File: pyanaconda/ui/gui/spokes/lib/resize.glade (hunk 2 above)
Comment (vojtechtrefny) [reply pyanaconda/ui/gui/spokes/lib/resize.glade:29]: Glade does this -- just opening and saving the file with a newer version will make some changes like these automatically. I guess the property names got changed between and 3.18 and 3.40...

File: pyanaconda/modules/storage/devicetree/viewer.py [hunk 3]
```diff
@@ -56,43 +56,43 @@ def get_root_device(self):
         :return: a name of the root device
```
Comment (rvykydal) [inline pyanaconda/modules/storage/devicetree/viewer.py:56]: I think the doc string should be updated.

File: pyanaconda/modules/storage/partitioning/manual/manual_module.py [hunk 4]
```diff
@@ -73,8 +76,11 @@ def setup_kickstart(self, data):
         for request in self.requests:
             mount_data = data.MountData()
             mount_data.mount_point = request.mount_point
-            mount_data.device = request.device_spec
+
+            device = self.storage.devicetree.get_device_by_device_id(request.device_spec)
+            mount_data.device = device.path
```
Comment (rvykydal) [inline pyanaconda/modules/storage/partitioning/manual/manual_module.py:81]: Just out of interest, could you explain why device.path was chosen to be used as device_spec, or add some context? (I guess it is available for all the devices while being unique enough ... ?) Does it have some caveats?

File: pyanaconda/ui/lib/storage.py [hunk 5]
```diff
@@ -376,11 +376,11 @@ def ignore_oemdrv_disks():
             disk_select_proxy.IgnoredDisks = ignored_disks
 
 
-def filter_disks_by_names(disks, names):
-    """Filter disks by the given names.
+def filter_disks_by_names(disks, disk_ids):
```
Comment (rvykydal) [inline pyanaconda/ui/lib/storage.py:379]: Should we update also name of the function?

File: tests/unit_tests/pyanaconda_tests/modules/storage/test_module_storage.py [hunk 6]
```diff
@@ -904,6 +909,7 @@ def test_mount_kickstart(self, publisher):
         # Mount points configuration
         mount /dev/sda1 /boot
         """
+        self._add_device(StorageDevice("sda1"))
```
Comment (rvykydal) [inline tests/unit_tests/pyanaconda_tests/modules/storage/test_module_storage.py:912]: I wonder why this became needed.

**Review by rvykydal — APPROVED** [review]
Looks good to me.

File: tests/unit_tests/pyanaconda_tests/modules/storage/test_module_storage.py (hunk 6 above)
Comment (vojtechtrefny) [reply tests/unit_tests/pyanaconda_tests/modules/storage/test_module_storage.py:912]: `MountPointRequest` now contains the `device_id` so the manual module needs to be able to resolve the device so we need it in the devicetree for the test.
